File: Back/nest-backend/scrape_ntc.py
import requests
from bs4 import BeautifulSoup
import json
import urllib3
import concurrent.futures
import time

urllib3.disable_warnings()

BASE_URL = "https://stat.ntc.tj/Y26/RPlan?page="
TOTAL_PAGES = 300

specialties = {}

def scrape_page(page):
    try:
        response = requests.get(BASE_URL + str(page), verify=False, timeout=10)
        soup = BeautifulSoup(response.text, 'html.parser')
        table = soup.find('table')
        if not table:
            return []
        
        rows = table.find_all('tr')
        page_specs = []
        for row in rows[1:]: # Skip header
            cols = row.find_all('td')
            if len(cols) >= 3:
                # Column 0: Code, Column 1: Specialty name (usually). We need to verify.
                # Actually, the structure might be: Муассиса, Кластер, Рамз, Ихтисос...
                # Let's extract text of all columns and find common patterns.
                row_data = [c.text.strip() for c in cols]
                page_specs.append(row_data)
        return page_specs
    except Exception as e:
        print(f"Error on page {page}: {e}")
        return []

print("Starting scraper...")
all_data = []

# Scrape first page to see the format before doing everything
test_data = scrape_page(1)
if test_data and len(test_data) > 0:
    print(f"Sample row length: {len(test_data[0])}")
    print(f"Sample row: {test_data[0]}")
    
    # Run concurrently for first 200 pages
    with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:
        futures = {executor.submit(scrape_page, p): p for p in range(1, 201)}
        for future in concurrent.futures.as_completed(futures):
            p = futures[future]
            try:
                res = future.result()
                all_data.extend(res)
                if p % 20 == 0:
                    print(f"Processed 20 pages... Total rows so far: {len(all_data)}")
            except Exception as exc:
                print('%r generated an exception: %s' % (p, exc))

    # We need to map it carefully now.
    # NTC Data typically: [Code, Specialty, University, Language, Plan, ...]
    # Let's just save the raw data in JSON and analyze it.
    with open('ntc_raw_data.json', 'w', encoding='utf-8') as f:
        json.dump(all_data, f, ensure_ascii=False, indent=2)
    print(f"Done. Scraped {len(all_data)} rows.")
else:
    print("Could not parse table.")

File: Back/nest-backend/scrape_ntc.js
const https = require('https');
const fs = require('fs');

const BASE_URL = 'https://stat.ntc.tj/Y26/RPlan?page=';

async function fetchPage(page, retries = 3) {
    for (let i = 0; i <= retries; i++) {
        try {
            return await new Promise((resolve, reject) => {
                const req = https.get(BASE_URL + page, { rejectUnauthorized: false, timeout: 30000 }, (res) => {
                    let data = '';
                    res.on('data', chunk => data += chunk);
                    res.on('end', () => resolve(data));
                });
                req.on('error', err => reject(err));
                req.on('timeout', () => {
                    req.destroy();
                    reject(new Error('Timeout'));
                });
            });
        } catch (err) {
            if (i === retries) throw err;
            console.log(`⚠️ Retrying page ${page} (${i + 1}/${retries})...`);
            await new Promise(r => setTimeout(r, 2000 * (i + 1))); // Exponential backoff
        }
    }
}

function parseHtml(pageHtml) {
    let rowsFound = 0;
    let pageData = [];
    let rows = pageHtml.split('<tr>');
    for (let i = 1; i < rows.length; i++) {
        let rowHtml = rows[i].split('</tr>')[0];
        let tds = rowHtml.split(/<td[^>]*>/);
        if (tds.length < 3) continue;
        
        let rowData = [];
        for (let j = 1; j < tds.length; j++) {
            let cellData = tds[j].split('</td>')[0].replace(/<[^>]+>/g, '').trim();
            rowData.push(cellData);
        }
        if (rowData.length > 0) {
            pageData.push(rowData);
            rowsFound++;
        }
    }
    return pageData;
}

async function scrape() {
    console.log('🚀 Starting Fast Batched Scraper...');
    let allData = [];
    const BATCH_SIZE = 5;
    const MAX_PAGES = 350;

    for (let p = 1; p <= MAX_PAGES; p += BATCH_SIZE) {
        let batchPromises = [];
        let end = Math.min(p + BATCH_SIZE - 1, MAX_PAGES);
        console.log(`📡 Fetching batch: pages ${p} to ${end}...`);

        for (let i = p; i <= end; i++) {
            batchPromises.push(fetchPage(i).catch(err => {
                console.error(`❌ Page ${i} failed:`, err.message);
                return null;
            }));
        }

        const results = await Promise.all(batchPromises);
        let batchRows = 0;
        
        for (let resHtml of results) {
            if (resHtml) {
                const parsed = parseHtml(resHtml);
                allData.push(...parsed);
                batchRows += parsed.length;
            }
        }

        console.log(`✅ Batch complete. Found ${batchRows} rows. Total so far: ${allData.length}`);
        
        if (batchRows === 0 && p > 100) {
            console.log("🛑 No rows found in this batch. Ending scrape early.");
            break;
        }
    }

    fs.writeFileSync('ntc_raw_data.json', JSON.stringify(allData, null, 2));
    console.log(`🏁 FINISHED! Scraped ${allData.length} records. Saved to ntc_raw_data.json`);
}

scrape();
